swarm/storage/pushsync/pusher_test.go

```diff
@@ -0,0 +1,414 @@
+			}
+		}()
+		return nil
+	}
+	// register the respond function
+	ps.Register(pssChunkTopic, false, respond)
+	tags, tagIDs := setupTags(chunkCnt, tagCnt)
+	// construct the mock push sync index iterator
+	tp := newTestPushSyncIndex(chunkCnt, tagIDs, tags, sent)
+	// start push syncing in a go routine
+	p := New(tp, ps, tags)
+	defer p.Close()
+	// collect synced chunks until all chunks synced
+	// wait on errc for errors on any thread
+	// otherwise time out
+	for {
+		select {
+		case i := <-tp.synced:
+			sent.Delete(i)
+			n := synced[i]
+			synced[i] = n + 1
+			if len(synced) == chunkCnt {
+				expTotal := chunkCnt / tagCnt
+				checkTags(t, expTotal, tagIDs[:tagCnt-1], tags)
+				return
+			}
+		case err := <-errc:
+			if err != nil {
+				t.Fatal(err)
+			}
+		case <-time.After(timeout):
+			t.Fatalf("timeout waiting for all chunks to be synced")
+		}
+	}
+
+}
+
+// setupTags constructs tags object create tagCnt - 1 tags
+// the sequential fake chunk i will be tagged with i%tagCnt
+func setupTags(chunkCnt, tagCnt int) (tags *chunk.Tags, tagIDs []uint32) {
+	// construct tags object
+	tags = chunk.NewTags()
+	// all but one tag is created
+	for i := 0; i < tagCnt-1; i++ {
+		tags.New("", int64(chunkCnt/tagCnt))
+	}
+	// extract tag ids
+	tags.Range(func(k, _ interface{}) bool {
+		tagIDs = append(tagIDs, k.(uint32))
+		return true
+	})
+	// add an extra for which no tag exists
+	return tags, append(tagIDs, 0)
+}
+
+func checkTags(t *testing.T, expTotal int, tagIDs []uint32, tags *chunk.Tags) {
+	for _, tagID := range tagIDs {
+		tag, err := tags.Get(tagID)
+		if err != nil {
+			t.Fatalf("expected no error getting tag '%v', got %v", tagID, err)
+		}
+		n, total, err := tag.Status(chunk.StateSent)
+		if err != nil {
+			t.Fatalf("getting status for tag '%v', expected no error, got %v", tagID, err)
+		}
+		if int(n) != expTotal {
+			t.Fatalf("expected Sent count on tag '%v' to be %v, got %v", tagID, expTotal, n)
+		}
+		if int(total) != expTotal {
+			t.Fatalf("expected Sent count on tag '%v' to be %v, got %v", tagID, expTotal, n)
+		}
+		n, total, err = tag.Status(chunk.StateSynced)
+		if err != nil {
+			t.Fatalf("getting status for tag '%v', expected no error, got %v", tagID, err)
+		}
+		if int(n) != expTotal {
+			t.Fatalf("expected Sent count on tag '%v' to be %v, got %v", tagID, expTotal, n)
+		}
+		if int(total) != expTotal {
+			t.Fatalf("expected Sent count on tag '%v' to be %v, got %v", tagID, expTotal, n)
+		}
+	}
+}
+
+type testStore struct {
+	store *sync.Map
+}
+
+func (t *testStore) Put(_ context.Context, _ chunk.ModePut, ch chunk.Chunk) (bool, error) {
+	cur := binary.BigEndian.Uint64(ch.Address()[:8])
+	var storedCnt uint32 = 1
+	v, loaded := t.store.LoadOrStore(cur, &storedCnt)
+	if loaded {
+		atomic.AddUint32(v.(*uint32), 1)
+	}
+	return false, nil
+}
+
+// TestPushSyncAndStoreWithLoopbackPubSub tests the push sync protocol
+// push syncer node communicate with storers via mock PubSub
```

Comment:
I would appreciate comments that describe the test procedure a bit more in detail.